graphbolt/src/csc_sampling_graph.cc

```diff
@@ -252,40 +249,55 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighborsImpl(
-                    torch::empty({allocate_size}, indptr_options);
-                torch::Tensor& picked_tensor =
-                    picked_neighbors_cur_thread[i - begin];
-                AT_DISPATCH_INTEGRAL_TYPES(
-                    picked_tensor.scalar_type(), "CallPick", ([&] {
-                      pick_fn(
-                          offset, num_neighbors,
-                          picked_tensor.data_ptr<scalar_t>());
-                    }));
-
-                num_picked_neighbors_per_node[i + 1] = allocate_size;
+                num_picked_neighbors_data_ptr[i + 1] =
+                    num_neighbors == 0 ? 0 : num_pick_fn(offset, num_neighbors);
               }
-              picked_neighbors_per_thread[thread_id] =
-                  torch::cat(picked_neighbors_cur_thread);
-            });  // End of parallel_for.
+            });
       }));
+
+  // Step 2. Calculate prefix sum to get total length and offsets of each node.
+  // It's also the indptr of the generated subgraph.
   torch::Tensor subgraph_indptr =
       torch::cumsum(num_picked_neighbors_per_node, 0);
 
-  torch::Tensor picked_eids = torch::cat(picked_neighbors_per_thread);
+  // Step 3. Allocate the tensor for picked neighbors.
+  torch::Tensor picked_eids;
+  AT_DISPATCH_INTEGRAL_TYPES(
+      subgraph_indptr.scalar_type(), "GetTotalLength", ([&] {
+        const auto total_length =
+            subgraph_indptr.data_ptr<scalar_t>()[num_nodes];
+        picked_eids = torch::empty({total_length}, indptr_options);
+      }));
+
+  // Step 4. Pick neighbors for each node.
+  AT_DISPATCH_INTEGRAL_TYPES(
+      picked_eids.scalar_type(), "GetPickedNeighbors", ([&] {
+        torch::parallel_for(
+            0, num_nodes, grain_size, [&](int64_t begin, int64_t end) {
+              const scalar_t* indptr_data = indptr_.data_ptr<scalar_t>();
+              auto picked_eids_data_ptr = picked_eids.data_ptr<scalar_t>();
+              auto num_picked_neighbors_data_ptr =
+                  num_picked_neighbors_per_node.data_ptr<scalar_t>();
+              auto subgraph_indptr_data_ptr =
+                  subgraph_indptr.data_ptr<scalar_t>();
+
+              const auto nodes_data_ptr = nodes.data_ptr<int64_t>();
+              for (int64_t i = begin; i < end; ++i) {
+                const auto nid = nodes_data_ptr[i];
+                const auto offset = indptr_data[nid];
+                const auto num_neighbors = indptr_data[nid + 1] - offset;
+                const auto picked_number = num_picked_neighbors_data_ptr[i + 1];
+                const auto picked_offset = subgraph_indptr_data_ptr[i];
+                if (picked_number > 0) {
+                  pick_fn(
+                      offset, num_neighbors,
+                      picked_eids_data_ptr + picked_offset);
+                }
+              }
+            });
+      }));
+
+  // Step 5. Calculate other attributes and return the subgraph.
   torch::Tensor subgraph_indices =
```

Comment:
We can consider pre-allocate the memory and move the `index_select` to multi-thread section if it's really slow after test. But this could be a future work.

Reply:
As I'm testing the performance, this `index_select` is sometimes the most time-consuming part and seems to be unstable in terms of performance. So I think it needs improvement.

Reply:
Moving this inside the multi-thread section would increase labor performance by a lot as labor already accesses the indices array while sampling.